runtime/fiber.c

```diff
@@ -0,0 +1,536 @@
+/*                                                                        */
+/*      KC Sivaramakrishnan, Indian Institute of Technology, Madras       */
+/*                   Tom Kelly, OCaml Labs Consultancy                    */
+/*                Stephen Dolan, University of Cambridge                  */
+/*                                                                        */
+/*   Copyright 2021 Indian Institute of Technology, Madras                */
+/*   Copyright 2021 OCaml Labs Consultancy                                */
+/*   Copyright 2019 University of Cambridge                               */
+/*                                                                        */
+/*   All rights reserved.  This file is distributed under the terms of    */
+/*   the GNU Lesser General Public License version 2.1, with the          */
+/*   special exception on linking described in the file LICENSE.          */
+/*                                                                        */
+/**************************************************************************/
+
+#define CAML_INTERNALS
+
+#include <string.h>
+#include <unistd.h>
+#include "caml/alloc.h"
+#include "caml/codefrag.h"
+#include "caml/fail.h"
+#include "caml/fiber.h"
+#include "caml/gc_ctrl.h"
+#include "caml/platform.h"
+#include "caml/minor_gc.h"
+#include "caml/misc.h"
+#include "caml/major_gc.h"
+#include "caml/memory.h"
+#include "caml/startup_aux.h"
+#ifdef NATIVE_CODE
+#include "caml/stack.h"
+#include "caml/frame_descriptors.h"
+#endif
+
+#ifdef DEBUG
+#define fiber_debug_log(...) caml_gc_log(__VA_ARGS__)
+#else
+#define fiber_debug_log(...)
+#endif
+
+void caml_change_max_stack_size (uintnat new_max_size)
+{
+  struct stack_info *current_stack = Caml_state->current_stack;
+  asize_t size = Stack_high(current_stack) - (value*)current_stack->sp
+                 + Stack_threshold / sizeof (value);
+
+  if (new_max_size < size) new_max_size = size;
+  if (new_max_size != caml_max_stack_size){
+    caml_gc_log ("Changing stack limit to %"
+                 ARCH_INTNAT_PRINTF_FORMAT "uk bytes",
+                     new_max_size * sizeof (value) / 1024);
+  }
+  caml_max_stack_size = new_max_size;
+}
+
+struct stack_info** caml_alloc_stack_cache () {
+  int i;
+
+  struct stack_info** stack_cache =
+    (struct stack_info**)caml_stat_alloc_noexc(sizeof(struct stack_info*) *
+                                               NUM_STACK_SIZE_CLASSES);
+  if (stack_cache == NULL)
+    return NULL;
+
+  for(i = 0; i < NUM_STACK_SIZE_CLASSES; i++)
+    stack_cache[i] = NULL;
+
+  return stack_cache;
+}
+
+Caml_inline struct stack_info* alloc_for_stack (mlsize_t wosize)
+{
+  return caml_stat_alloc_noexc(sizeof(struct stack_info) +
+                               sizeof(value) * wosize +
+                               8 /* for alignment */ +
+                               sizeof(struct stack_handler));
```

Comment:
For OpenBSD support, we're going to need to allocate this stack memory with `mmap(..., MAP_STACK|MAP_ANON|MAP_PRIVATE)` or else the process is killed by the syscall entry function. I've got a patch for this locally on OpenBSD but am not sure of the performance impact of switching to mmap for every OS or if we should just do that for OpenBSD.

Reply:
Tracking this for post-merge fixing in https://github.com/ocaml/ocaml/issues/10861